File: test-suite/golden-tests/uint8-cross-language/ops.R
## R implementations for the uint8-cross-language golden test.
##
## The morloc type [UInt8] maps to R's `raw` here (declared in main.loc
## as `type R => (List UInt8) = "raw" UInt8`). The R runtime always
## materialises a [UInt8] wire packet as a RAWSXP regardless of hint, so
## the type assertions below should always hold. If a future runtime
## regression delivers something else (e.g. an `integer` vector or a
## `list`), the echo and native functions stop with a clear message
## that the test will capture.

r_encode <- function(s) {
    charToRaw(s)
}

r_decode <- function(b) {
    if (typeof(b) != "raw") {
        stop(paste0("r_decode expected raw, got ", typeof(b)))
    }
    rawToChar(b)
}

r_echo <- function(b) {
    if (typeof(b) != "raw") {
        stop(paste0("r_echo expected raw, got ", typeof(b)))
    }
    b
}

r_native <- function(b) {
    paste0("r: ", typeof(b), ", length=", length(b))
}

File: test-suite/golden-tests/uint8-cross-language/ops.py
"""
Python implementations for the uint8-cross-language golden test.

The morloc type [UInt8] maps to Python `bytes` here (via root-py's
`type Py => (List UInt8) = "bytes" UInt8` specialisation). Every
function below either produces a `bytes` value or asserts that its
input is a `bytes` value -- if the runtime ever delivers a `list`
instead (e.g. because of a wire-deserialiser regression), the echo
and native functions raise a TypeError that the test will surface
through obs.err.
"""


def py_encode(s):
    return s.encode('utf-8')


def py_decode(b):
    if not isinstance(b, bytes):
        raise TypeError(
            "py_decode expected bytes (the contracted native form of "
            "[UInt8] in Python), got " + type(b).__name__
        )
    return b.decode('utf-8')


def py_echo(b):
    if not isinstance(b, bytes):
        raise TypeError(
            "py_echo expected bytes, got " + type(b).__name__
        )
    return b


def py_native(b):
    return "python: " + type(b).__name__ + ", len=" + str(len(b))
